### PR TITLE
feat: add user API for settings/contact-info

### LINKED ISSUE #504: Implement user API on settings/contact info
**I'm submitting a ...**  (check one with "x")
- [ ] bug report
- [x] feature request
- [ ] support request => Please do not submit support requests here, instead ask your query in out Gitter channel at https://gitter.im/fossasia/open-event-frontend

screenshot from 2017-07-19 14-07-15

### PR DESCRIPTION
**Pull request by sumedh123**
#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

Fixes #504

### CODE REVIEW

**Comment by sumedh123** [comment]
@niranjan94 I implemented the users api for this route and I am posting the correct data, the server side shows the correct data but I am getting error in console on the client side. 'Unknow error'.
Following is my request payload on the server side:
payloadd
Following is the error I am getting:
contacterror

**Comment by niranjan94** [comment]
Try to make a request manually using api and see first..

**Comment by sumedh123** [comment]
@niranjan94 I tried to send the patch request manually also with the same access token as I get from my dev tools for a user, it's showing the same error.
patch
patch2
Please guide.

**Comment by niranjan94** [comment]
@sumedh123 please open an issue on the server repo with endpoint, request + body, response body.

**Comment by sumedh123** [comment]
@niranjan94 Please review. https://oefrontend2017.herokuapp.com

File: app/controllers/settings/contact-info.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import Ember from 'ember';
+
+const { Controller } = Ember;
+export default Controller.extend({
+  actions: {
+    updateContactInfo() {
+      this.set('isLoading', true);
```
Comment (CosmicCoder96) [inline app/controllers/settings/contact-info.js]: where are you using this in the template?

**Comment by sumedh123** [comment]
@niranjan94 @CosmicCoder96 @geekyd Please review. https://oefrontend2017.herokuapp.com

File: app/controllers/settings/contact-info.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import Ember from 'ember';
+
+const { Controller } = Ember;
+
+export default Controller.extend({
+  actions: {
+    updateContactInfo() {
+      this.set('isLoading', true);
+      let currentUser = this.get('model');
+      currentUser.save({
```
Comment (niranjan94) [inline app/controllers/settings/contact-info.js]: Remove all arguments

File: app/serializers/user.js [hunk 3]
```diff
@@ -5,9 +5,13 @@ export default ApplicationSerializer.extend({
   serialize(snapshot, options) {
     const json = this._super(...arguments);
     if (snapshot.id) {
-      let attributesToOmit = ['email'];
+      let attributesToOmit = [];
       if (!snapshot.adapterOptions || !snapshot.adapterOptions.includePassword) {
         attributesToOmit.push('password');
+      } else if (!snapshot.adapterOptions || !snapshot.adapterOptions.includeEmail) {
```
Comment (niranjan94) [inline app/serializers/user.js]: Revert this change. Not needed.

File: tests/integration/components/settings/contact-info-test.js [hunk 4]
```diff
@@ -4,7 +4,13 @@ import hbs from 'htmlbars-inline-precompile';
 
 moduleForComponent('settings/contact-info-section', 'Integration | Component | settings/contact info section');
 
+var user = {
```
Comment (niranjan94) [inline tests/integration/components/settings/contact-info-test.js]: Should be an ember object

File: app/serializers/user.js (hunk 3 above)
Comment (sumedh123) [inline app/serializers/user.js]: @niranjan94 Just the else if part of 'includeEmail'?

**Comment by niranjan94** [comment]
Only the first if condition for includePassword is needed. The rest to be
removed.

File: app/serializers/user.js (hunk 3 above)
Comment (sumedh123) [inline app/serializers/user.js]: @niranjan94 If we just keep the first condition ovf includePassword, how would we pick the required fields? It throws error saying 'Missing data for required field.' since all other attributes will also be there in the request body

Comment (niranjan94) [inline app/serializers/user.js]: You don't have to pick attributes. As long as the rest are null, the API should accept. We only ensure there is no password field. All the other fields can be there.

Just see how the profile page works. https://github.com/fossasia/open-event-frontend/pull/512 (The email was ignored there before because the API had an issue with email. now since that is fixed, it need not be ignored anymore.)

Comment (sumedh123) [inline app/serializers/user.js]: @niranjan94 But the thing is api is not accepting. It is demanding the email and password field.
email-api4

Comment (niranjan94) [inline app/serializers/user.js]: @sumedh123 yes. Password is already omitted. And email is already going in. I donno what you're trying to do here.

Comment (niranjan94) [inline app/serializers/user.js]: Again, just see profile page.

Comment (sumedh123) [inline app/serializers/user.js]: @niranjan94 I saw that. I implemented like that only. What is happening is by the following code:
```
import ApplicationSerializer from 'open-event-frontend/serializers/application';
import { pick, omit } from 'lodash';

export default ApplicationSerializer.extend({
  serialize(snapshot, options) {
    const json = this._super(...arguments);
    if (snapshot.id) {
      let attributesToOmit = [];
      if (!snapshot.adapterOptions || !snapshot.adapterOptions.includePassword) {
        attributesToOmit.push('password');
      }
      json.data.attributes = omit(json.data.attributes, attributesToOmit);
      console.log(json.data.attributes);
    } else if (options && options.includeId) {
      json.data.attributes = pick(json.data.attributes, ['email', 'password']);
    }
    return json;
  }
});
```
password is getting ommitted. Still console shows me the error that 'Missing field required' for password.
email-api8

Comment (niranjan94) [inline app/serializers/user.js]: Why are you making a POST request @sumedh123 ? POST is for register. The docs cannot get any clearer than this. 
- https://open-event-api.herokuapp.com/#users-user-details-patch 
- https://guides.emberjs.com/v2.14.0/models/creating-updating-and-deleting-records/#toc_persisting-records 

Thats the reason the API is expecting a password and an email. I donno how you're making a request here. save should be called on an existing user for ember to do a PATCH request.

Comment (niranjan94) [inline app/serializers/user.js]: Saying it once again ... Profile page does exactly what is needed.

Comment (sumedh123) [inline app/serializers/user.js]: @niranjan94 Pushed the code. I am using the save method from the beginning itself and following the same as the profile page. Please guide if I am wrong. Also I tried logging at this place:
```
else if (options && options.includeId) {
      console.log('post');
      json.data.attributes = pick(json.data.attributes, ['email', 'password']);
    }
```
but 'post' is not getting logged. So I guess it should be patch.

File: app/controllers/settings/contact-info.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+export default Controller.extend({
+  actions: {
+    updateContactInfo() {
+      this.set('isLoading', true);
+      let currentUser = this.get('model');
+      currentUser.save()
+        .then(user => {
+          this.set('isLoading', false);
+          let userData = user.serialize(false).data.attributes;
```
Comment (niranjan94) [inline app/controllers/settings/contact-info.js]: Remove lines 13,14,15. This will be done globally later.

File: app/controllers/settings/contact-info.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+          userData.id = user.get('id');
+          this.get('session').set('data.currentUserFallback', userData);
+          this.get('notify', 'Updated information successfully');
+        })
+        .catch(error => {
+          this.set('isLoading', false);
+          this.get('authManager.currentUser').rollbackAttributes();
```
Comment (niranjan94) [inline app/controllers/settings/contact-info.js]: Instead of doing it here, run `this.get('authManager.currentUser').rollbackAttributes();` when the user exits the route.

**Comment by sumedh123** [comment]
@niranjan94 I pushed the code. Please review. I am still seeing the POST request in my console though. https://oefrontend2017.herokuapp.com

**Comment by niranjan94** [comment]
@sumedh123 how is it that you're seeing a `POST` request ?

screen shot 2017-07-23 at 3 26 00 pm

What exactly are you doing ?

**Comment by niranjan94** [comment]
Also please fix your tests

**Comment by sumedh123** [comment]
@niranjan94 I fixed it. Pushing the code.

**Comment by sumedh123** [comment]
@niranjan94 Please review. https://oefrontend2017.herokuapp.com

**Comment by sumedh123** [comment]
@niranjan94 Please review.

File: app/controllers/settings/contact-info.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+export default Controller.extend({
+  actions: {
+    updateContactInfo() {
+      this.set('isLoading', true);
+      let currentUser = this.get('model');
+      currentUser.save()
+        .then(() => {
+          this.set('isLoading', false);
+          this.get('notify', 'Updated information successfully');
```
Comment (CosmicCoder96) [inline app/controllers/settings/contact-info.js]: @sumedh123 I don't think this is the right way to display the success notification, also it's not working either. No notification appears on deployment link after updation.

**Comment by sumedh123** [comment]
@CosmicCoder96 Please review now. I fixed it. https://oefrontend2017.herokuapp.com

**Comment by sumedh123** [comment]
@geekyd @CosmicCoder96 Please merge this if no other changes are needed.
